### PR TITLE
Add Junit5 JVM annotations

### PR DESCRIPTION
**Pull request by loicmathieu**
This is a draft PR to gather feeback on JVM annotations with JUnit 5.
It contains some code around forking that still needs improvements.
It also provides disabling forking wich should be discussed.

This is buid on top of PR #22 that needs to be merged first.

### CODE REVIEW

**Comment by loicmathieu** [comment]
@jeanbisutti with these changes tests are failing on java8 and java9 saying that this version of Java didn't include JFR. I don't understand how it is possible ... can you have a look and help me on this ?

```
Expecting:
 <"Java Flight Recorder profiling is available for 
Oracle JDK >= 1.7u40 and OpenJDK >= 11">
to contain:
 <"Rule: Primitive To Object Conversion"> 
```

**Comment by jeanbisutti** [comment]
There is a bug in the display of this message. I think that there is an exception when the test class annotated with @QuickPerf test is executed.

**Comment by loicmathieu** [comment]
@jeanbisutti I reproduce it on localhost with Oracle Java 8.
So there is definilty a bug somewhere that prevents JFR annotations to works on JUnit 5 with Java 8 and 9.

**Comment by jeanbisutti** [comment]
@loicmathieu The known bug is about the message displayed, not about the JFR recording. An exception not related to JFR could explain this message.

**Comment by jeanbisutti** [comment]
The full stack trace may give insight of what happens.

**Comment by loicmathieu** [comment]
@jeanbisutti after sabling the JFR test everything works so I'm pretty sure it's JFR that didn't works

**Comment by loicmathieu** [comment]
Full stack trace, nothing shows up :(
```
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.077 s <<< FAILURE! - in org.quickperf.junit5.jvm.jmc.JmcTests
[ERROR] jmc_rules_having_score_greater_than_expected  Time elapsed: 0.077 s  <<< FAILURE!
org.assertj.core.api.SoftAssertionError: 

The following 2 assertions failed:
1) 
Expecting:
 <"Java Flight Recorder profiling is available for 
Oracle JDK >= 1.7u40 and OpenJDK >= 11">
to contain:
 <"JMC rules are expected to have score less than <50>."> 
2) 
Expecting:
 <"Java Flight Recorder profiling is available for 
Oracle JDK >= 1.7u40 and OpenJDK >= 11">
to contain:
 <"Rule: Primitive To Object Conversion"> 

	at org.quickperf.junit5.jvm.jmc.JmcTests.jmc_rules_having_score_greater_than_expected(JmcTests.java:82)
```

**Comment by jeanbisutti** [comment]
@loicmathieu More information is required to understand what happens. JFR recording was extensively tested with JUnit 4 and several JDK.

**Comment by jeanbisutti** [comment]
@loicmathieu, not this stack trace. Thz stack trace of the test class annotated with @QuickPerfTest, as mentioned above.

**Comment by loicmathieu** [comment]
> @loicmathieu, not this stack trace. Thz stack trace of the test class annotated with @QuickPerfTest, as mentioned above.

This is the full stack trace available ...

**Comment by jeanbisutti** [comment]
Hi @loicmathieu 
Thanks for the time you spent on QuickPerf development.
I have just investigated from your code.
First of all,  `JavaFlightRecorderProfilerBeforeJava9 ` class is used to start and stop JFR profiling before Java 9. From Java 9, `JavaFlightRecorderProfilerFromJava9` is used. In addition, before Java 11, JFR profiling requires `-XX:+UnlockCommercialFeatures` JVM option (see  `JfrJvmOptions `). 

 In `QuickPerfExtension`, I noticed the following code:
```java
       if (testExecutionContext.testExecutionUsesTwoJVMs() && !forkDisabled && "false".equals(inAFork)) {
            exePerfInstrumentBeforeTestMethod(perfRecordersToExecuteBeforeTestMethod);
            AllJvmOptions jvmOptions = testExecutionContext.getJvmOptions();
            newJvmTestLauncher.run( invocationContext.getExecutable()
                    , testExecutionContext.getWorkingFolder()
                    , jvmOptions
                    , QuickPerfJunit5Core.class);
            exePerfInstrumentAfterTestMethod(perfRecordersToExecuteAfterTestMethod);
```
The run method of NewJvmTestLauncher was not executed. So I executed the test with the following if condition to try to have the `-XX:+UnlockCommercialFeatures` JVM option.
```java
       if (testExecutionContext.testExecutionUsesTwoJVMs()
```
It did not work but I then noticed that the error you reported appeared before invoking the run method of NewJvmTestLauncher. The test method body and the perf recorders should not be executed in the main JVM (launching the test JVM).
I hope that these elements will help.

**Comment by jeanbisutti** [comment]
@loicmathieu It si possible that the extension mechanism of JUnit 5 has some limitations to make JVM annotations work. In such a case, a custom [TestEngine](https://junit.org/junit5/docs/current/api/org/junit/platform/engine/TestEngine.html) (perhaps extending JupiterTestEngine) could make JVM annotations work.

**Comment by loicmathieu** [comment]
@jeanbisutti it will help :) I now understand what's going on. Will try to fix this next week.

**Comment by jeanbisutti** [comment]
Thanks a lot for your work @loicmathieu ! :)

**Comment by loicmathieu** [comment]
@jeanbisutti I make some progress :)

The recorder was not triggered on the forked JVM as you suspected. After correcting this, I now have a huge byte allocation start offset difference. I'm not sure it's a real issue.
I will try to find some time to ends this PR soon ...

**Comment by jeanbisutti** [comment]
Great @loicmathieu! :) 
Regarding the greater allocation offset with JUnit 5, it might be caused by the allocation of `org.junit.platform.commons.util.ReflectionUtils.invokeMethod`.
You may have forgotten to commit the new allocation offset for JUnit 5. Interested to see the CI results with this offset.

**Comment by loicmathieu** [comment]
@jeanbisutti I now successfully have almost all test pass except the Oracle Java 8 ...
The implementation is not elegant but it works :)

For Java 8 tests, it fails on a strange error ...

**Comment by loicmathieu** [comment]
@jeanbisutti it should works now. I have a silent exception for parsing Java version ...
Now, all the offsets are OK for all version of Java supported.

**Comment by jeanbisutti** [comment]
Great @loicmathieu! :) 
The code can be clean up in a second step.
In `retrieveJunitByteOffset` mehod of `ByteWatcherRecorder`, each time a `JvmVersion` method is called, it allocates on heap, and this allocation has to be deduced after... 
If `retrieveJunitByteOffset` method does not allocate, the implementation should be simpler.
With JUnit 4, the fact that `retrieveJunitByteOffset` allocates leads to some allocation tests to fail in some environments (some people trying to build QuickPerf complained about that but now it seems ok).
However, it may have another implementation even simpler and more stable. I confess that I am a little worried about the high allocation offset with JUnit 5. The tests might not be stable in different environements. For JUnit 4, you have the following code in `QuickPerfMethod` class:
```java
public Object invokeExplosively(final Object target, final Object... params)
            throws Throwable {
        return new ReflectiveCallable() {
            @Override
            protected Object runReflectiveCall() throws Throwable {
                startRecordings();
                try {
                    return method.invoke(target, params);
                } finally {
                    stopRecordings();
                }
            }
        }.run();
    }
```
In `interceptTestMethod` method of `QuickPerfTestExtension`, I suppose that the `proceed` method of `MethodInvocation` is called:
```
public T proceed() {
        return ReflectionUtils.invokeMethod(this.method, this.target.orElse((Object)null), this.arguments);
    }
```
The invocation of this method might be responsible for the high allocation offset required with JUnit 5.
Maybe, as for JUnit 4, the `QuickPerfTestExtension` could also called `method.invoke(target, params)` but we should before investigate the reason of `ReflectionUtils.invokeMethod`...

**Comment by jeanbisutti** [comment]
@loicmathieu 
I have an idea to maintain the compatibilty with JUnit 5, to have a code simpler and to have tests stable in different environments.
In `ReflectionUtils`, we have the following code:
```java
 public static Object invokeMethod(Method method, Object target, Object... args) {
        Preconditions.notNull(method, "Method must not be null");
        Preconditions.condition(target != null || isStatic((Member)method), () -> {
            return String.format("Cannot invoke non-static method [%s] on a null target.", method.toGenericString());
        });

        try {
            return ((Method)makeAccessible(method)).invoke(target, args);
        } catch (Throwable var4) {
            throw ExceptionUtils.throwAsUncheckedException(getUnderlyingCause(var4));
        }
    }
```
The idea is to call all the code other than `invoke(target, args);`  before the invocation of `startRecording(perfRecordersToExecuteBeforeTestMethod);` in `QuickPerfExtension`.

**Comment by jeanbisutti** [comment]
@loicmathieu 
The test method may be invoked twice with JUnit 5 and a JVM annotation.
With a `System.out` added in the test method, two lines are printed in the console. By adding `Thread.sleep`, `Thread.sleep` seems to be executed twice. That could be an issue if the user wants to do a performance benchmark with a test during several minutes.

**Comment by loicmathieu** [comment]
@jeanbisutti yes, now, this is mandatory for the moment as JUnit doesn't allow to not proceed an invocation. I openned an issue on Junit for this: https://github.com/junit-team/junit5/issues/2083

I hope they will allow it to be implemented ...

**Comment by loicmathieu** [comment]
> The idea is to call all the code other than `invoke(target, args)` before the invocation of `startRecording(perfRecordersToExecuteBeforeTestMethod)` in `QuickPerfExtension`.

@jeanbisutti what you suggest is to bypass the processing of the invocation in the forked JVM and to run the same code as JUnit ReflectionUtils ? 
For the moment it is not possible as the invocation must be proceed or JUnit will fail ...

**Comment by loicmathieu** [comment]
> The invocation of this method might be responsible for the high allocation offset required with JUnit 5.
> Maybe, as for JUnit 4, the `QuickPerfTestExtension` could also called `method.invoke(target, params)` but we should before investigate the reason of `ReflectionUtils.invokeMethod`...

 `ReflectionUtils.invokeMethod` if to check the precondition of the method, as they are already executed in the host VM maybe we can re-work the fork to directly call the target method.

As it is a lot of works, can you merge the PR as is? It is still better than the current support for JUnit5 :)

**Comment by loicmathieu** [comment]
@jeanbisutti I try to refactor the allocation byteoffset a little, it didn't creates any allocation on itself but JUnit5 offsets are still high.

**Comment by loicmathieu** [comment]
@jeanbisutti the build fail to launch Java 13 build, can you re-launch it ?
I think that all should be OK now

**Comment by jeanbisutti** [comment]
Thanks a lot for your work @loicmathieu! The build is re-launched. Regarding the other issues, I suggest to talk face to face at next Devoxx Belgium, it will be more efficient.

**Comment by loicmathieu** [comment]
@jeanbisutti I update the code to what we discussed.
CI is happy with it :)

I think this is the best we can do while waiting for JUnit 5.6 release that should allow to skip the invocation and improving the code (I put a bunch of FIXME for this).

For me, it's ready do go !

File: junit5/src/main/java/org/quickperf/junit5/QuickPerfTestExtension.java [hunk 1]
```diff
@@ -52,20 +59,59 @@ public void interceptTestMethod(Invocation<Void> invocation,
...
+                businessThrowable = throwable;
             }
+            finally {
+                if(!testExecutionContext.isQuickPerfDisabled()){
+                    stopRecording(perfRecordersToExecuteAfterTestMethod);
+                }
+            }
+        }
+
+        //FIXME why didn't the ConsoleReporter nor JUnit shows the throwable ?
```
Comment (jeanbisutti) [inline junit5/src/main/java/org/quickperf/junit5/QuickPerfTestExtension.java]: It is OK: issueThrower.throwIfNecessary(businessThrowable, groupOfPerfIssuesToFormat);
I created a PR to check this with an automatic test

File: jvm-annotations/src/main/java/org/quickperf/jvm/allocation/bytewatcher/ByteWatcherRecorder.java [hunk 2]
```diff
@@ -23,29 +23,67 @@
...
+    private static int junit4ByteOffset = 40;
+    private static int java12ByteOffsetForJunit4 = 72;
+
+    private static int junit5ByteOffset = 13_728;
+    private static int java9ByteOffsetForJunit5 =  12_120;
+    private static int java10ByteOffsetForJunit5 =  11_744;
+    private static int java12ByteOffsetForJunit5 = 11_296;
+    private static int java13ByteOffsetForJunit5 = 11_264;
+
+    //determine junit4 and junit5 offset statically to avoid as much allocation as possible when calling startRecording
```
Comment (jeanbisutti) [inline jvm-annotations/src/main/java/org/quickperf/jvm/allocation/bytewatcher/ByteWatcherRecorder.java]: Could you please remove the word statically

**Comment by jeanbisutti** [comment]
@loicmathieu Great! :)

I am a little afraid by the huge allocation offset. Based on JUnit 4 experience, some tests may fail in other environments.

In QuickPerfExtension, the following code may reduce the JUnit 5 offset:
```java
if(SystemProperties.TEST_CODE_EXECUTING_IN_NEW_JVM.evaluate()) {

            Method method = invocationContext.getExecutable();
            Object target = invocationContext.getTarget().get();
            List<Object> args = invocationContext.getArguments();

            if (!testExecutionContext.isQuickPerfDisabled()) {
                startRecording(perfRecordersToExecuteBeforeTestMethod);
            }
            try {
                //invocation.proceed();
                method.invoke(target, args);
                //FIXME call directly the test, this needs to allow to skip the invocation (see bellow)
                //invocation.skip();
            } finally {
                if (!testExecutionContext.isQuickPerfDisabled()) {
                    stopRecording(perfRecordersToExecuteAfterTestMethod);
                }
            }
            return;
        }
```

**Comment by loicmathieu** [comment]
> In QuickPerfExtension, the following code may reduce the JUnit 5 offset

As we discuss in Devoxx, this is not currently possible as the invocation must be proceed or JUnit will throws an error.
When JUnit 5.6 will be released, we will be able to reduce a lot the allocation offset by some similar techniques. I have a fix for this but it need to wait for the next release of JUnit.

**Comment by jeanbisutti** [comment]
The build could be unstable because of the high allocation offset with JUnit 5. It is something very annoying in my opinion.

Waiting for JUnit 5.6 release, we could use the following hack to have a low allocation offset with JUnit 5 and to avoid having a "Chain of InvocationInterceptors never called invocation":
```java
        Field field = invocation.getClass().getDeclaredField("invoked");
        field.setAccessible(true);

        AtomicBoolean invoked = (AtomicBoolean) field.get(invocation);
        invoked.set(true);
```
Another option could be to wait for JUnit 5.6.

**Comment by loicmathieu** [comment]
@jeanbisutti I prefere not to incude such hack for only a few weeks ... As JUnit5 module is experimental I prefere to keep it that way and update it as soon as JUnit 5.6 will be out.
Maybe we need a simple way to disabled the allocation test (using a maven profile ?) because even with lower offset the issue still arise (I don't have the sale offset on my machine than the one used by the CI).
I can raise an issue to provides a way to skip the allocation test for both JUnit5 and JUnit4 if you want

**Comment by jeanbisutti** [comment]
@loicmathieu With JUnit 4, allocation tests seem regularly ok in the CI. In the past, as far as I know, users having an allocation offset issue with JUnit 4 told me it works fine now. Do you have an allocation offset issue with JUnit 4?

**Comment by loicmathieu** [comment]
@jeanbisutti I don't have an allocation issue with JUnit4 but the noth implementation are based on measured allocation offset that can depends on JVM implementation, OS, hardware, ... So I think that both are innacurate.

**Comment by jeanbisutti** [comment]
@loicmathieu Some tests will fail if allocation measure is one byte greater or lower than expected. I notice that the allocation tests are regularly green in the CI for several JDK (Oracle JDK, OpenJDK JDK, from JDK 1.8 to  JDK 13). CI builds uses Linux apart from one build using OS X. For Windows, I tested from JDK 1.7 to JDK 13 (Oracle JDK and OpenJDK JDK). A friend tested with several JDK on Mac. Users having an issue with JUnit 4 told me it works fine now. Today, based on the above informations, it seems that the allocation measures work quite well with Oracle JDK and OpenJDK JDK.

**Comment by loicmathieu** [comment]
@jeanbisutti I found why the exception stacktrace was not reporting, I forget to print the failures ...
I add a commit that fixes this.

File: jvm-annotations/src/main/java/org/quickperf/jvm/allocation/bytewatcher/ByteWatcherRecorder.java (hunk 2 above)
Comment (loicmathieu) [reply jvm-annotations/src/main/java/org/quickperf/jvm/allocation/bytewatcher/ByteWatcherRecorder.java]: done

File: junit5/src/main/java/org/quickperf/junit5/QuickPerfTestExtension.java (hunk 1 above)
Comment (loicmathieu) [reply junit5/src/main/java/org/quickperf/junit5/QuickPerfTestExtension.java]: This was a bug, I forget to print the failure report.
This is fixed now.

**Comment by loicmathieu** [comment]
@jeanbisutti I revert the hack to skip the invocation as the invocation hierarchy is make of composable invocation. We cannot be 100% sure that the invocation inside the interceptor is a `ValidatingInvocation` (it can be an other type of Invocation that wrap a `ValidatingInvocation`), so we cannot blindly swicth it's `invoked` field.

So for now, there is two possibilities:

- Merge it as is, maybe disabling the allocation test to avoid too much issue with it. if so we should clrealy states in the documentation that the module is experimental and the test is launched twice for some annotations
- Wait for JUnit5 5.6 but I don't know when it will be avaiable (can be weeks or months ...)

I let you choose :)

**Comment by jeanbisutti** [comment]
Thanks for your work @loicmathieu! Another solution could be to temporarly use  JUnit 5.6 - SNAPHOT

**Comment by loicmathieu** [comment]
@jeanbisutti I finally succeeded in skipping the invocation and makes all this works by using JUnit 5.6.0-SNAPSHOT.
I still need to fix all the offsets, so I will push multiple times in order the CI to help me on this :)

**Comment by loicmathieu** [comment]
@jeanbisutti CI is green, allocation offsets are low, ready to merge !

**Comment by jeanbisutti** [comment]
Huge thanks for your work @loicmathieu !
Could you please apply the minor changes and squash the commits into one?
After that, I will merge the commit.
The ByteWatcherRecorder should not be aware of Junit 4 and JUnit 5.  I will refactor this after the merge. This refactoring will be useful for the TestNG version.

**Comment by loicmathieu** [comment]
@jeanbisutti I squashed all commits into one.
I don't know which minor changes you talked about, I think I address all the changes you requested ?

File: junit5-jvm-test/src/test/java/org/quickperf/junit5/jvm/allocation/AllocationAnnotationsTest.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    private TestExecutionSummary testResult(Class<?> testClass) {
+        SummaryGeneratingListener listener = new SummaryGeneratingListener();
+        LauncherDiscoveryRequest request = LauncherDiscoveryRequestBuilder.request()
+                .selectors(selectClass(testClass))
+                .build();
+        Launcher launcher = LauncherFactory.create();
+        launcher.registerTestExecutionListeners(listener);
+        launcher.execute(request);
+        TestExecutionSummary  summary = listener.getSummary();
+        summary.printTo(new PrintWriter(System.out));
```
[the remark this answers is not in the document]
Comment (loicmathieu) [reply junit5-jvm-test/src/test/java/org/quickperf/junit5/jvm/allocation/AllocationAnnotationsTest.java]: I found them useful to debug quickperf integration with JUnit issue, that's why I keep them in the test. It's a test so it's harmless

**Comment by loicmathieu** [comment]
@jeanbisutti I take all your remarks into accounts.

**Pull request merged**